src/controller/java/src/chip/setuppayload/SetupPayloadParser.kt

```diff
@@ -0,0 +1,108 @@
+   * @param allowInvalidPayload Allow invalid payload components. If this value is true, payload
+   * element validation is not checked. Consider saying that the payload must still parse
+   * correctly, but this skips validation of the content past parsing (i.e. it does not validate
+   * ranges for individual elements). Refer to [SetupPayload] for the description of the
+   * invalid element.
+   */
+  @Throws(UnrecognizedQrCodeException::class, SetupPayloadException::class)
+  fun parseQrCode(qrCodeString: String, allowInvalidPayload: Boolean): SetupPayload {
+    return fetchPayloadFromQrCode(qrCodeString, allowInvalidPayload)
+  }
+
+  /**
+   * Returns [SetupPayload] parsed from the manual entry code string. If an SetupPINCode has
+   * invalid value, SetupPayloadException occurs. Refer to [SetupPayload] for the description
+   * of the invalid element.
+   */
+  @Throws(InvalidEntryCodeFormatException::class, SetupPayloadException::class)
+  fun parseManualEntryCode(entryCodeString: String): SetupPayload {
+    return fetchPayloadFromManualEntryCode(entryCodeString, false)
+  }
+
+  /**
+   * Returns [SetupPayload] parsed from the manual entry code string.
+   *
+   * @param entryCodeString the manual Pairing Code for commissioning device.
+   * @param allowInvalidPayload Allow invalid payload components. If this value is true, payload
+   * element validation is not checked. Consider saying that the payload must still parse
+   * correctly, but this skips validation of the content past parsing (i.e. it does not validate
+   * ranges for individual elements). Refer to [SetupPayload] for the description of the
+   * invalid element.
+   */
+  @Throws(InvalidEntryCodeFormatException::class, SetupPayloadException::class)
+  fun parseManualEntryCode(entryCodeString: String, allowInvalidPayload: Boolean): SetupPayload {
+    return fetchPayloadFromManualEntryCode(entryCodeString, allowInvalidPayload)
+  }
+
+  /** Get QR code string from [SetupPayload].  */
+  @Throws(SetupPayloadException::class)
+  external fun getQrCodeFromPayload(payload: SetupPayload): String?
+
+  /** Get manual entry code string from [SetupPayload].  */
+  @Throws(SetupPayloadException::class)
+  external fun getManualEntryCodeFromPayload(payload: SetupPayload): String?
+
+  @Throws(UnrecognizedQrCodeException::class, SetupPayloadException::class)
+  private external fun fetchPayloadFromQrCode(
+    qrCodeString: String, isAllowInvalidPayload: Boolean
+  ): SetupPayload
+
+  @Throws(InvalidEntryCodeFormatException::class, SetupPayloadException::class)
+  private external fun fetchPayloadFromManualEntryCode(
+    entryCodeString: String, isAllowInvalidPayload: Boolean
+  ): SetupPayload
+
+  class UnrecognizedQrCodeException(qrCode: String) :
+    Exception(String.format("Invalid QR code string: %s", qrCode), null) {
+    companion object {
+      private const val serialVersionUID = 1L
+    }
+  }
+
+  class InvalidEntryCodeFormatException(entryCode: String) :
```

Comment:
"Manual pairing code" is the name of in the spec. "Entry code" is new terminology.